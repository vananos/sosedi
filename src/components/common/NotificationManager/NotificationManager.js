import React, { Component } from "react";
import Notification from "../Notification/Notification";
import "./NotificationManager.scss";

export default class NotificationManager extends Component {
  static notificationIdCounter = 0;

  constructor(props) {
    super(props);
    this.state = {
      notifications: []
    };
    NotificationManager._singletonRef = this;
  }

  static notify = (msg, { duration = 3000, type = "success" } = {}) => {
    const notificationData = {
      type,
      msg,
      key: NotificationManager.notificationIdCounter++
    };

    const hideFunction = () =>
      NotificationManager._singletonRef.deleteNotification(notificationData);
    notificationData.onClick = hideFunction;

    NotificationManager._singletonRef.addNotification(notificationData);

    setInterval(hideFunction, duration);
  };

  addNotification = notifcation =>
    this.setState({
      notifications: [...this.state.notifications, notifcation]
    });

  deleteNotification = notifcationForDelete => {
    const filteredNotifications = this.state.notifications.filter(
      notifcation => notifcation !== notifcationForDelete
    );
    this.setState({ notifications: filteredNotifications });
  };

  render() {
    return (
      <div className="notification-area">
        {this.state.notifications.map(notifcation => (
          <Notification {...notifcation} />
        ))}
      </div>
    );
  }
}
