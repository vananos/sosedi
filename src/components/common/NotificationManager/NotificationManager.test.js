import React from "react";
import NotificationManager from "./NotificationManager";
import { shallow } from "enzyme";

const notificationContent = <span>notification content</span>;

describe("NotificationManager, ", () => {
  let notificationManagerWrapper;

  beforeEach(() => {
    notificationManagerWrapper = shallow(<NotificationManager />);
  });

  it("when notify is called -> notification should appear", () => {
    NotificationManager.notify(notificationContent);

    expect(notificationManagerWrapper).toMatchSnapshot();
  });

  it("when notification lifetime is over -> notification should be removed", async () => {
    NotificationManager.notify(notificationContent, { duration: 500 });
    await delay(600);
    expect(notificationManagerWrapper).toMatchSnapshot();
  });
});

const delay = async duration => {
  return new Promise((resolve, reject) =>
    setTimeout(() => resolve("done"), duration)
  );
};
