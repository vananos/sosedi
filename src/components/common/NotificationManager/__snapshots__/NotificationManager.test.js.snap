// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationManager,  when notification lifetime is over -> notification should be removed 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NotificationManager />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [],
      "className": "notification-area",
    },
    "ref": null,
    "rendered": Array [],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [],
        "className": "notification-area",
      },
      "ref": null,
      "rendered": Array [],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`NotificationManager,  when notify is called -> notification should appear 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NotificationManager />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Notification
          msg={
            <span>
              notification content
            </span>
          }
          onClick={[Function]}
          type="success"
        />,
      ],
      "className": "notification-area",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "function",
        "props": Object {
          "msg": <span>
            notification content
          </span>,
          "onClick": [Function],
          "type": "success",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Notification
            msg={
              <span>
                notification content
              </span>
            }
            onClick={[Function]}
            type="success"
          />,
        ],
        "className": "notification-area",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "function",
          "props": Object {
            "msg": <span>
              notification content
            </span>,
            "onClick": [Function],
            "type": "success",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
